compiler/rustc_llvm/llvm-wrapper/PassWrapper.cpp

```diff
@@ -55,6 +55,9 @@ typedef struct LLVMOpaqueTargetMachine *LLVMTargetMachineRef;
 DEFINE_STDCXX_CONVERSION_FUNCTIONS(Pass, LLVMPassRef)
 DEFINE_STDCXX_CONVERSION_FUNCTIONS(TargetMachine, LLVMTargetMachineRef)
 
+static std::string SavedCompilerPath;
+static std::vector<std::string> SavedOtherArgs;
```

Comment:
The problem is that I need to convert the strings into c++ std::strings and make sure their lifetime is not shorter than TargetMachine which contains them by reference.

To ensure the lifetime does not run out I added those as global. I put the initialisation of those values in a functions, which should only be called once, so I thought this is fine.
See 'LLVMRustSaveInvocationArguments' which is called by llvm_utils::configure_llvm which is called by llvm_utils::init. The init function ensures it is only called once.

The main problem is that they must be std::strings. Maybe I could create a function which creates an C++ object that has those strings, which can then be passed as parameter to LLVMRustCreateTargetMachine which must have a static lifetime?

Reply:
I created a new functions which creates args, that I'll pass to LLVMRustCreateTargetMachine. While I added a dispose function for the args, I don't really know how  and when I'd call it. So I'll essentially leak something everytime a new factory function is created, which happens when ModuleLlvm::new is called.
Fortunately I was able to avoid leaking everytime it is called, so hopefully its not too bad.

So the issue of clobbering global state should be fixed, at the cost of leaking everytime ModuleLlvm::new is called.

Reply:
Ok this happens more often then I thought, will be trying to find a way without leaking.